File: ios/joinery-kit/Sources/JoineryCalendarKit/CalendarAPI.swift
import Foundation
import JoineryKit

/// What the editor sends for a repeating entry.
public struct CalRecurrenceInput: Equatable, Sendable {
    public enum Ends: Equatable, Sendable {
        case never
        case onDate(String)     // "yyyy-MM-dd"
        case afterCount(Int)
    }

    public var type: String         // daily | weekly | monthly | yearly
    public var interval: Int
    public var daysOfWeek: [Int]    // weekly: 0=Sun…6=Sat
    public var weekOfMonth: Int?    // monthly by-weekday: 1-4, -1 = last
    public var ends: Ends

    public init(type: String, interval: Int = 1, daysOfWeek: [Int] = [], weekOfMonth: Int? = nil, ends: Ends = .never) {
        self.type = type
        self.interval = interval
        self.daysOfWeek = daysOfWeek
        self.weekOfMonth = weekOfMonth
        self.ends = ends
    }

    var jsonValue: JSONValue {
        var body: [(key: String, value: JSONValue)] = [
            (key: "type", value: .string(type)),
            (key: "interval", value: .number(Double(max(1, interval)))),
        ]
        if !daysOfWeek.isEmpty {
            body.append((key: "days_of_week", value: .array(daysOfWeek.map { .number(Double($0)) })))
        }
        if let weekOfMonth {
            body.append((key: "week_of_month", value: .number(Double(weekOfMonth))))
        }
        switch ends {
        case .never:
            body.append((key: "ends", value: .string("never")))
        case .onDate(let date):
            body.append((key: "ends", value: .string("date")))
            body.append((key: "end_date", value: .string(date)))
        case .afterCount(let count):
            body.append((key: "ends", value: .string("count")))
            body.append((key: "count", value: .number(Double(count))))
        }
        return .object(body)
    }
}

/// Thin typed face over the core calendar actions (calendar_feed,
/// calendar_entry, calendar_entry_save, calendar_entry_delete). Ownership is
/// enforced server-side; every call rides the app's session key.
public struct CalendarAPI: Sendable {
    let client: APIClient

    public init(client: APIClient) {
        self.client = client
    }

    public func feed(startUTC: String, endUTC: String) async throws -> CalendarFeed {
        let envelope = try await client.submitAction("calendar_feed", body: .object([
            (key: "start", value: .string(startUTC)),
            (key: "end", value: .string(endUTC)),
        ]))
        guard let feed = CalendarFeed(data: envelope["data"]) else {
            throw JoineryAPIError.malformedResponse
        }
        return feed
    }

    public func entry(id: Int) async throws -> CalEntryDetail {
        let envelope = try await client.submitAction("calendar_entry", body: .object([
            (key: "entry_id", value: .number(Double(id))),
        ]))
        guard let detail = CalEntryDetail(data: envelope["data"]) else {
            throw JoineryAPIError.malformedResponse
        }
        return detail
    }

    /// Create or update. `scope` + `occurrenceDate` drive the recurring
    /// series semantics exactly as on the web (this / future / all).
    @discardableResult
    public func save(
        entryID: Int?,
        occurrenceDate: String?,
        scope: String?,
        date: String,
        title: String,
        allDay: Bool,
        startTime: String?,
        endTime: String?,
        blocks: Bool,
        timezone: String,
        recurrence: CalRecurrenceInput?
    ) async throws -> Int {
        var body: [(key: String, value: JSONValue)] = [
            (key: "date", value: .string(date)),
            (key: "title", value: .string(title)),
            (key: "all_day", value: .bool(allDay)),
            (key: "blocks", value: .bool(blocks)),
            (key: "timezone", value: .string(timezone)),
        ]
        if let entryID {
            body.append((key: "entry_id", value: .number(Double(entryID))))
        }
        if let occurrenceDate {
            body.append((key: "occurrence_date", value: .string(occurrenceDate)))
        }
        if let scope {
            body.append((key: "scope", value: .string(scope)))
        }
        if !allDay {
            body.append((key: "start_time", value: .string(startTime ?? "")))
            body.append((key: "end_time", value: .string(endTime ?? "")))
        }
        if let recurrence {
            body.append((key: "recurrence", value: recurrence.jsonValue))
        }
        let envelope = try await client.submitAction("calendar_entry_save", body: .object(body))
        return envelope["data"]?["entry_id"]?.intValue ?? entryID ?? 0
    }

    public func delete(entryID: Int, scope: String?, occurrenceDate: String?) async throws {
        var body: [(key: String, value: JSONValue)] = [
            (key: "entry_id", value: .number(Double(entryID))),
        ]
        if let scope {
            body.append((key: "scope", value: .string(scope)))
        }
        if let occurrenceDate {
            body.append((key: "occurrence_date", value: .string(occurrenceDate)))
        }
        _ = try await client.submitAction("calendar_entry_delete", body: .object(body))
    }
}

File: android/joinery-android-calendar/src/main/java/com/getjoinery/calendar/CalendarApi.kt
package com.getjoinery.calendar

import com.getjoinery.android.ApiClient
import com.getjoinery.android.JoineryApiError
import com.getjoinery.android.JsonValue

/** What the editor sends for a repeating entry. */
data class CalRecurrenceInput(
    val type: String,           // daily | weekly | monthly | yearly
    val interval: Int = 1,
    val daysOfWeek: List<Int> = emptyList(),  // weekly: 0=Sun…6=Sat
    val weekOfMonth: Int? = null,             // monthly by-weekday: 1-4, -1 = last
    val ends: Ends = Ends.Never,
) {
    sealed class Ends {
        object Never : Ends()
        data class OnDate(val date: String) : Ends()   // "yyyy-MM-dd"
        data class AfterCount(val count: Int) : Ends()
    }

    fun jsonValue(): JsonValue {
        val body = ArrayList<Pair<String, JsonValue>>()
        body.add("type" to JsonValue.Str(type))
        body.add("interval" to JsonValue.Num(maxOf(1, interval).toDouble()))
        if (daysOfWeek.isNotEmpty()) {
            body.add("days_of_week" to JsonValue.Arr(daysOfWeek.map { JsonValue.Num(it.toDouble()) }))
        }
        if (weekOfMonth != null) {
            body.add("week_of_month" to JsonValue.Num(weekOfMonth.toDouble()))
        }
        when (ends) {
            is Ends.Never -> body.add("ends" to JsonValue.Str("never"))
            is Ends.OnDate -> {
                body.add("ends" to JsonValue.Str("date"))
                body.add("end_date" to JsonValue.Str(ends.date))
            }
            is Ends.AfterCount -> {
                body.add("ends" to JsonValue.Str("count"))
                body.add("count" to JsonValue.Num(ends.count.toDouble()))
            }
        }
        return JsonValue.Obj(body)
    }
}

/**
 * Thin typed face over the core calendar actions (calendar_feed,
 * calendar_entry, calendar_entry_save, calendar_entry_delete). Ownership is
 * enforced server-side; every call rides the app's session key.
 */
class CalendarApi(val client: ApiClient) {

    suspend fun feed(startUtc: String, endUtc: String): CalendarFeed {
        val envelope = client.submitAction(
            "calendar_feed",
            JsonValue.obj("start" to JsonValue.Str(startUtc), "end" to JsonValue.Str(endUtc)),
        )
        return CalendarFeed.from(envelope["data"]) ?: throw JoineryApiError.Malformed
    }

    suspend fun entry(id: Int): CalEntryDetail {
        val envelope = client.submitAction(
            "calendar_entry",
            JsonValue.obj("entry_id" to JsonValue.Num(id.toDouble())),
        )
        return CalEntryDetail.from(envelope["data"]) ?: throw JoineryApiError.Malformed
    }

    /** Create or update. [scope] + [occurrenceDate] drive the recurring
     *  series semantics exactly as on the web (this / future / all). */
    suspend fun save(
        entryId: Int?,
        occurrenceDate: String?,
        scope: String?,
        date: String,
        title: String,
        allDay: Boolean,
        startTime: String?,
        endTime: String?,
        blocks: Boolean,
        timezone: String,
        recurrence: CalRecurrenceInput?,
    ): Int {
        val body = ArrayList<Pair<String, JsonValue>>()
        body.add("date" to JsonValue.Str(date))
        body.add("title" to JsonValue.Str(title))
        body.add("all_day" to JsonValue.Bool(allDay))
        body.add("blocks" to JsonValue.Bool(blocks))
        body.add("timezone" to JsonValue.Str(timezone))
        if (entryId != null) body.add("entry_id" to JsonValue.Num(entryId.toDouble()))
        if (occurrenceDate != null) body.add("occurrence_date" to JsonValue.Str(occurrenceDate))
        if (scope != null) body.add("scope" to JsonValue.Str(scope))
        if (!allDay) {
            body.add("start_time" to JsonValue.Str(startTime ?: ""))
            body.add("end_time" to JsonValue.Str(endTime ?: ""))
        }
        if (recurrence != null) body.add("recurrence" to recurrence.jsonValue())
        val envelope = client.submitAction("calendar_entry_save", JsonValue.Obj(body))
        return envelope["data"]?.get("entry_id")?.intValue ?: entryId ?: 0
    }

    suspend fun delete(entryId: Int, scope: String?, occurrenceDate: String?) {
        val body = ArrayList<Pair<String, JsonValue>>()
        body.add("entry_id" to JsonValue.Num(entryId.toDouble()))
        if (scope != null) body.add("scope" to JsonValue.Str(scope))
        if (occurrenceDate != null) body.add("occurrence_date" to JsonValue.Str(occurrenceDate))
        client.submitAction("calendar_entry_delete", JsonValue.Obj(body))
    }
}
